src/lib.rs

```diff
@@ -430,6 +429,61 @@ where
 
         SieveHandle::new(handle, disp_rx)
     }
+
+    async fn disconnect<SA: Sampler>(&self, peer: PublicKey, sender: Arc<S>, sampler: Arc<SA>) {
+        if self.gossip.read().await.contains(&peer) {
+            debug!("peer {} from our gossip set disconnected", peer);
+
+            let mut gossip = self.gossip.write().await;
+
+            gossip.remove(&peer);
+
+            let not_gossip = sender
+                .keys()
+                .await
+                .into_iter()
+                .filter(|x| !gossip.contains(&x));
+
+            if let Ok(new) = sampler.sample(not_gossip, 1).await {
+                debug!("resampled for {} new peers", new.len());
+
+                gossip.extend(new);
+            }
+        }
+
+        let sender = Arc::new(ConvertSender::new(sender));
+
+        self.murmur.disconnect(peer, sender, sampler).await;
+    }
+
+    async fn garbage_collection(&self) {
+        let mut batches = self.pending.write().await;
+        let mut delivered = self.delivered.write().await;
+
+        // FIXME: this should use `HashMap::drain_filter` once it is stabilized
+        let expired_digests = batches
+            .iter()
+            .filter_map(|(digest, batch)| {
+                if batch.is_expired(self.config.expiration_delay()) {
+                    Some(*digest)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for digest in expired_digests {
+            batches.remove(&digest);
+            delivered.remove(&digest);
+            self.echoes.purge(digest).await;
+        }
+
+        // yes this is ugly, blame the type inferer
```

Comment:
bad, bad inferer! :laughing:

Reply:
Even more when the notation to specifiy trait generics is so ugly